File: Professional/ActiveReportsWebPro/C#/ActiveReportsWebPro/RpxHandler.cs
using System;
using System.IO;
using System.Web;
using System.Web.Caching;
using GrapeCity.ActiveReports.Export.Html;
namespace GrapeCity.ActiveReports.Samples.Web.Professional
{
	public class RpxHandler : IHttpHandler
	{
		public const string HandlerCacheExtension = ".rpxWeb";
		public void ProcessRequest(HttpContext context)
		{
			// return image
			var keyName = Path.GetFileName(context.Request.FilePath);
			var cacheItem = context.Cache[keyName];
			context.Response.BinaryWrite((byte[])cacheItem);
		}
		public bool IsReusable
		{
			get { return true; }
		}
		public sealed class HtmlOutputHandler : IOutputHtml
		{
			private readonly Cache _cache;
			private readonly string _name;
			public HtmlOutputHandler(Cache cache, string name)
			{
				_cache = cache;
				_name = name;
			}
			public string OutputHtmlData(HtmlOutputInfoArgs info)
			{
				var stream = info.OutputStream;
				var dataLen = (int)stream.Length;
				if (dataLen <= 0)
					return string.Empty;
				var bytesData = new byte[dataLen];
				stream.Seek(0, SeekOrigin.Begin);
				stream.Read(bytesData, 0, dataLen);
				if (info.OutputKind == HtmlOutputKind.HtmlPage)
				{
					MainPageData = bytesData;
					return _name;
				}
				var keyName = Guid.NewGuid() + HandlerCacheExtension;
				// 30 seconds should be enough to load all images
				_cache.Insert(keyName, bytesData, null, Cache.NoAbsoluteExpiration, new TimeSpan(0, 0, 30));
				return keyName;
			}
			public void Finish()
			{
			}
			public byte[] MainPageData { get; private set; }
		}
	}
}

File: Professional/ActiveReportsWebPro/VB.NET/ActiveReportsWebPro/RpxHandler.vb
Imports System.IO
Imports System.Web
Imports System.Web.Caching
Imports GrapeCity.ActiveReports.Export.Html
Public Class RpxHandler
	Implements IHttpHandler
	Private Const HandlerCacheExtension As String = ".rpxWeb"
	Public Sub ProcessRequest(context As HttpContext) Implements IHttpHandler.ProcessRequest
		' return image
		Dim keyName As String = Path.GetFileName(context.Request.FilePath)
		Dim cacheItem = DirectCast(context.Cache(keyName), Byte())
		context.Response.BinaryWrite(cacheItem)
	End Sub
	Public ReadOnly Property IsReusable As Boolean Implements IHttpHandler.IsReusable
		Get
			Return True
		End Get
	End Property
	Public NotInheritable Class HtmlOutputHandler
		Implements IOutputHtml
		Private ReadOnly _cache As Cache
		Private ReadOnly _name As String
		Private _mainPageData As Byte()
		Public Sub New(cache As Cache, name As String)
			_cache = cache
			_name = name
		End Sub
		Public Function OutputHtmlData(info As HtmlOutputInfoArgs) As String Implements IOutputHtml.OutputHtmlData
			Dim stream As Stream = info.OutputStream
			Dim dataLen = CInt(stream.Length)
			If dataLen <= 0 Then
				Return String.Empty
			End If
			Dim bytesData = New Byte(dataLen - 1) {}
			stream.Seek(0, SeekOrigin.Begin)
			stream.Read(bytesData, 0, dataLen)
			If info.OutputKind = HtmlOutputKind.HtmlPage Then
				MainPageData = bytesData
				Return _name
			End If
			Dim keyName As String = Guid.NewGuid().ToString() & HandlerCacheExtension
			' 30 seconds should be enough to load all images
			_cache.Insert(keyName, bytesData, Nothing, Cache.NoAbsoluteExpiration, New TimeSpan(0, 0, 30))
			Return keyName
		End Function
		Public Sub Finish() Implements IOutputHtml.Finish
		End Sub
		Public Property MainPageData As Byte()
			Get
				Return _mainPageData
			End Get
			Private Set(value As Byte())
				_mainPageData = value
			End Set
		End Property
	End Class
End Class
